File: python/sgf-parsing/sgf_parsing.py
class SgfTree(object):
    def __init__(self, properties=None, children=None):
        self.properties = properties or {}
        self.children = children or []

    def __eq__(self, other):
        if not isinstance(other, SgfTree):
            return False
        for k, v in self.properties.items():
            if k not in other.properties:
                return False
            if other.properties[k] != v:
                return False
        for k in other.properties.keys():
            if k not in self.properties:
                return False
        if len(self.children) != len(other.children):
            return False
        for a, b in zip(self.children, other.children):
            if a != b:
                return False
        return True


def parse(input_string):
    input_string = input_string.replace('\n', ' ').replace('\t', ' ')
    if not (input_string.startswith('(') and input_string.endswith(')')):
        raise ValueError('Must enclosed with parentheses!')

    s = input_string[1:-1]
    if not s.startswith(';'):
        raise ValueError('Tree with no nodes!')

    s = s[1:]

    if not s:
        return SgfTree()

    begin_index = find_index(s, '[')
    if begin_index < 0:
        raise ValueError('Properties without delimiter!')

    property_id = s[:begin_index]
    if not property_id.isupper():
        raise ValueError('Property id must be uppercase!')

    properties = {property_id: []}
    while begin_index < len(s) and s[begin_index] == '[':
        end_index = find_index(s, ']', begin_index)
        properties[property_id].append(decode(s[begin_index + 1:end_index]))

        begin_index = end_index + 1

    s = s[begin_index:]

    children = []
    index = 0
    while index < len(s):
        if s[index] == ';':
            end_index = find_index(s, ';(', index + 1)
            if end_index >= 0:
                end_index -= 1
            else:
                end_index = len(s) - 1

            children.append(parse('(' + s[index:end_index + 1] + ')'))
        elif s[index] == '(':
            end_index = find_index(s, ')', index)
            children.append(parse(s[index:end_index + 1]))

        index = end_index + 1

    return SgfTree(properties, children)


def find_index(s, target, offset=0):
    index = offset
    while index < len(s):
        if s[index] in target:
            return index

        if s[index] == '\\':
            index += 2
        else:
            index += 1
    return -1


def decode(s):
    return s.replace('\\', '')

File: java/sgf-parsing/src/main/java/SgfParsing.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SgfParsing {
  public SgfNode parse(String input) throws SgfParsingException {
    if (!(input.startsWith("(") && input.endsWith(")"))) {
      throw new SgfParsingException("Must enclosed with parentheses!");
    }

    String s = input.substring(1, input.length() - 1);
    if (!s.startsWith(";")) {
      throw new SgfParsingException("Tree with no nodes!");
    }

    s = s.substring(1);

    if (s.isEmpty()) {
      return new SgfNode();
    }

    if (!s.contains("[")) {
      throw new SgfParsingException("Properties without delimiter!");
    }

    Map<String, List<String>> properties = new HashMap<>();
    while (!s.isEmpty() && ";(".indexOf(s.charAt(0)) == -1) {
      int beginIndex = find(s, "[", 0);
      String propertyId = s.substring(0, beginIndex);
      if (propertyId.chars().anyMatch(c -> !Character.isUpperCase(c))) {
        throw new SgfParsingException("Property id must be uppercase!");
      }

      properties.put(propertyId, new ArrayList<>());
      while (beginIndex < s.length() && s.charAt(beginIndex) == '[') {
        int endIndex = find(s, "]", beginIndex);
        properties.get(propertyId).add(escape(s.substring(beginIndex + 1, endIndex)));

        beginIndex = endIndex + 1;
      }

      s = s.substring(beginIndex);
    }

    List<SgfNode> children = new ArrayList<>();
    int index = 0;
    while (index < s.length()) {
      int endIndex;
      if (s.charAt(index) == ';') {
        endIndex = find(s, ";(", index + 1);
        endIndex = ((endIndex == -1) ? s.length() : endIndex) - 1;

        children.add(parse("(" + s.substring(index, endIndex + 1) + ")"));
      } else {
        endIndex = find(s, ")", index);

        children.add(parse(s.substring(index, endIndex + 1)));
      }

      index = endIndex + 1;
    }

    return new SgfNode(properties, children);
  }

  String escape(String s) {
    StringBuilder result = new StringBuilder();
    int index = 0;
    while (index < s.length()) {
      if (s.startsWith("\\\\", index)) {
        result.append("\\");
        index += 2;
      } else if (s.startsWith("\t", index)) {
        result.append(" ");
        ++index;
      } else if (s.startsWith("\\\t", index)) {
        result.append(" ");
        index += 2;
      } else if (s.startsWith("\\\n", index)) {
        index += 2;
      } else if (s.startsWith("\\", index)) {
        ++index;
      } else {
        result.append(s.charAt(index));
        ++index;
      }
    }

    return result.toString();
  }

  int find(String s, String target, int offset) {
    int index = offset;
    while (index < s.length()) {
      if (target.indexOf(s.charAt(index)) != -1) {
        return index;
      }

      index += s.startsWith("\\", index) ? 2 : 1;
    }

    return -1;
  }
}
